File: 2015/06.py

import math, copy

with open('6_input') as f:
	data = f.read()
	data = data.split('\n')

light_field = [[0 for j in range(1000)] for i in range(1000)]

def part_1():
	for cmd in data:
		c = cmd.split()

		if c[0] == 'toggle':
			[start, _, end] = c[1:]
			[x1, y1] = list(map(int, start.split(',')))
			[x2, y2] = list(map(int, end.split(',')))

			for i in range(x1, x2 + 1):
				for j in range(y1, y2 + 1):
					light_field[i][j] += 2

		elif c[0] == 'turn':
			if c[1] == 'on':
				[start, _, end] = c[2:]
				[x1, y1] = list(map(int, start.split(',')))
				[x2, y2] = list(map(int, end.split(',')))
				for i in range(x1, x2 + 1):
					for j in range(y1, y2 + 1):
						light_field[i][j] += 1
				
			elif c[1] == 'off':
				[start, _, end] = c[2:]
				[x1, y1] = list(map(int, start.split(',')))
				[x2, y2] = list(map(int, end.split(',')))
				for i in range(x1, x2 + 1):
					for j in range(y1, y2 + 1):
						light_field[i][j] -= 1 if light_field[i][j] > 0 else 0
	lights_on = 0
	for i in range(1000):
		for j in range(1000):
			if light_field[i][j]:
				lights_on += light_field[i][j]
	print(lights_on)
	return

def part_2():
	return 

	data = data.split()
	data = list(map(int, data.split()))


if __name__ == '__main__':
	part_1()
	#part_2()
	

File: 2024/06.js

const fs = require('fs');
const { exec } = require("child_process");
const { ALPHA_L, ALPHA_U, NUMS, D4, D6, D8, MOD } = require("../lib");
const { joseph, findCycle } = require("../lib");
const { cache } = require("../lib");
const { bin, float, hex, int, num, oct } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { copy,	entries, in_, inBB, keys, sort, values } = require("../lib");
const { Counter, DD, empty, FastQueue, PriorityQueue, Queue, set, Stack } = require("../lib");
const { areaInt, circumference, manDist, shoelace } = require("../lib");
const { ord, chr, count, debug, disp, disp3, crt, gcd, lcm, modPow, modPowBig, modInv, mod, prod, prodBig, randint, sum, sumBig, transpose } = require("../lib");
const { digits, ints,	floats,	singles,	words,	lines,	table,	groups,	getGroups,	groupsWith,	parse,	parseLine, scanf } = require ('../lib');
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { disjoint, isSubset, isSuperset, or, and, xor, sub } = set;
const { getExecStr } = require("../lib/post");
const { combinations, combinations_with_replacement, next_permutation, product } = require("../lib");

const right = (dy, dx) => {
	return [dx, -dy];
}

function part1(data) {

	let res = 0;
	let sy, sx;
	data = table(data);
	iter(data, (r, i) => {
		iter(r, (c, j) => {
			if (c === '^') {
				sy = i;
				sx = j;
			}
		})
	})
	data[sy][sx] = '.'
	let [y, x] = [sy, sx];
	let dy = -1;
	let dx = 0;
	const visited = new set();
	let py, px;
	visited.add([y, x])
	while (true) {
		if (!inBB(y+dy, x+dx, data)) {
			break;
		}
		if (data[y+dy][x+dx] === '#') {
			[dy, dx] = right(dy, dx);
		}
		turn = 0;
		[py, px] = [y, x];
		[y, x] = [y+dy, x+dx];
		visited.add([y, x]);
	}
	res = visited.size;

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART1");
	part2(data, sy, sx, visited);
	return;
}

function part2(data, sy, sx, V) {

	let res = 0;
	iter(V, ([i, j]) => {
		if (i === sy && j === sx) return;
		data[i][j] = '#';
		let [y, x] = [sy, sx];
		let dy = -1;
		let dx = 0;
		const visited = new set();
		let isLoop = true
		let same = 0;
		while (true) {
			if (visited.has([y, x, dy, dx])) {
				break
			}
			visited.add([y, x, dy, dx]);
			if (!inBB(y+dy, x+dx, data)) {
				isLoop = false;
				break;
			}
			if (data[y+dy][x+dx] === '#') {
				[dy, dx] = right(dy, dx);
			} else {
				[y, x] = [y+dy, x+dx];
			}
		}

		if (isLoop) {
			++res;
		}

		data[i][j] = '.';
	})

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("06_input").toString("utf-8");

	part1(data);
	// part2(data);
	process.exit(0);
}

main();
